### PR TITLE
Switch RP2040 flash settings to nvm.toml

### PR DESCRIPTION
**Pull request by tannewt**
It generates a best case stage2 for all the possible flashes and will work with others of the same capability as well. Flash size is dynamically configured based on the flash.

This also moves TCM designated code and data into RAM so that it's space is freed in the XIP cache.

### CODE REVIEW

**Comment by ladyada** [comment]
want me to test feather rp2040 artifact? i could verify it uses QSPI?

File: ports/raspberrypi/Makefile [hunk 1]
```diff
@@ -270,9 +270,15 @@ $(BUILD)/boot2.bin: $(BUILD)/boot2.elf
 	$(STEPECHO) "OBJCOPY $<"
 	$(Q)$(OBJCOPY) -O binary $< $@
 
-$(BUILD)/boot2.elf: $(BOOT2_S_UPPER) | $(BUILD)/
```
Comment (jepler) [inline ports/raspberrypi/Makefile:273]: `BOOT2_S_UPPER ?= sdk/src/rp2_common/boot_stage2/boot2_generic_03h.S` and related lines should be removed unless it's possible/intended for boards to go back to an asm boot2 in which case removing this line is probably wrong

**Review by jepler — CHANGES_REQUESTED** [review]
A minor cleanup .. no testing performed.

File: ports/raspberrypi/Makefile (hunk 1 above)
Comment (jepler) [reply ports/raspberrypi/Makefile:273]: (other lines being the BOOT2_S_UPPER settings in individual board files)

File: ports/raspberrypi/gen_stage2.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import sys
+import cascadetoml
+import pathlib
+import typer
+from jinja2 import Template
+
+
+def main(input_template: pathlib.Path, output_path: pathlib.Path, skus: str = typer.Argument("")):
+    if ", " in skus:
+        skus = skus.split(", ")
```
Comment (dhalbert) [inline ports/raspberrypi/gen_stage2.py:10]: There are one or two board files (not in `raspberrypi`) that don't use `, `, but just `,`. How about testing for `,` and then split and trim, or use `re.split()`.

File: ports/raspberrypi/stage2.c.jinja [hunk 3]
```diff
@@ -0,0 +1,197 @@
+#include "sdk/src/rp2040/hardware_structs/include/hardware/structs/ssi.h"
```
Comment (dhalbert) [inline ports/raspberrypi/stage2.c.jinja:1]: I'm confused; I thought this file didn't fit in the 256 bytes, or did you remove some stuff or put it elsewhere?

File: ports/raspberrypi/supervisor/internal_flash.c [hunk 4]
```diff
@@ -51,25 +51,36 @@
...
         BINARY_INFO_BLOCK_DEV_FLAG_PT_UNKNOWN));
+
+    // Read the RDID register to get the flash capacity.
+    uint8_t cmd[] = {0x9f, 0, 0, 0};
+    uint8_t data[4];
+    flash_do_cmd(cmd, data, 4);
+    uint8_t power_of_two = 21;
+    if (data[3] >= 20 && data[3] < 30) {
+        power_of_two = data[3];
+    }
```
Comment (dhalbert) [inline ports/raspberrypi/supervisor/internal_flash.c:74]: Are there cases where `data[3]` is out of range (too small, probably)? It looks like you just assume 2MB (2**21)? Could you add a comment explaing this

File: ports/raspberrypi/supervisor/port.c [hunk 5]
```diff
@@ -74,10 +74,42 @@ STATIC void _binary_info(void) {
...
+extern uint32_t _ld_itcm_destination;
+extern uint32_t _ld_itcm_size;
+extern uint32_t _ld_itcm_flash_copy;
+
 safe_mode_t port_init(void) {
     _binary_info();
     // Set brown out.
 
+    // Copy all of the "tightly coupled memory" code and data to run from RAM.
+    // This let's us use the 16k cache for dynamically used data and code.
```
Comment (dhalbert) [inline ports/raspberrypi/supervisor/port.c:91]:
```suggestion
    // This lets us use the 16k cache for dynamically used data and code.
```

**Review by dhalbert — CHANGES_REQUESTED** [review]
I don't know some of technical details, but you are testing!

File: ports/raspberrypi/stage2.c.jinja (hunk 3 above)
Comment (tannewt) [reply ports/raspberrypi/stage2.c.jinja:1]: Removed the watchdog header.

**Comment by tannewt** [comment]
Ok, @jepler and @dhalbert, this is passing. Please take another look.

**Review by jepler — APPROVED** [review]
Thanks for resolving the issue I raised.

**Review by dhalbert — APPROVED** [review]
Looks good! This is a really good step for flash management.

**Pull request merged**